### PR TITLE
RavenDB-13712: Fix race in AsyncHiLoIdGenerator

### PR DESCRIPTION
**Pull request by garayx**

### CODE REVIEW

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs [hunk 1]
```diff
@@ -83,25 +83,30 @@ public async Task<long> NextIdAsync()
...
+                var nextTask = Interlocked.CompareExchange(ref _nextRangeTask, maybeNextTask, current);
+                if (nextTask == current) // replace was successful
                 {
-                    Interlocked.CompareExchange(ref _nextRangeTask, null, nextTask);
+                    await maybeNextTask.Value.ConfigureAwait(false);
+                    continue;
                 }
+
+                // failed to replace, let's wait on the previous task
+                await nextTask.Value.ConfigureAwait(false);
```
Comment (arekpalinski) [inline src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: You dropped `try .. finally` and effectively `_nextRangeTask` is no longer set to `null` on exception

Comment (garayx) [reply src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: done

**Review by ayende — APPROVED** [review]

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs [hunk 2]
```diff
@@ -83,25 +83,37 @@ public async Task<long> NextIdAsync()
...
                 {
-                    await nextTask.Value.ConfigureAwait(false);
+                    // let's try to call the existing task for next range
+                    await current.Value.ConfigureAwait(false);
+                    if (range != Range)
+                        continue;
+                }
+                catch
+                {
+                    // previous task was faulted, we will replace it
```
Comment (arekpalinski) [inline src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: How do you assure that we'll replace it? If someone replace `_nextRangeTask` meanwhile (possible, right?)  then below condition `if (nextTask == current)` will be false an we'll throw on:

```
// failed to replace, let's wait on the previous task
                await nextTask.Value.ConfigureAwait(false);
```

Comment (ayende) [reply src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: Good point, so we need to only throw an error higher IFF our own created task failed

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs [hunk 3]
```diff
@@ -83,25 +83,37 @@ public async Task<long> NextIdAsync()
...
+                var nextTask = Interlocked.CompareExchange(ref _nextRangeTask, maybeNextTask, current);
+                if (nextTask == current) // replace was successful
                 {
-                    Interlocked.CompareExchange(ref _nextRangeTask, null, nextTask);
+                    await maybeNextTask.Value.ConfigureAwait(false);
+                    continue;
                 }
+
+                // failed to replace, let's wait on the previous task
+                await nextTask.Value.ConfigureAwait(false);
```
Comment (ayende) [inline src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: This should be wrapped in try/catch and ignored.
Someone else failed, but we'll retry until our own task failed

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs (hunk 3 above)
Comment (garayx) [reply src/Raven.Client/Documents/Identity/AsyncHiLoIdGenerator.cs]: done

**Review by arekpalinski — APPROVED** [review]

**Review by ayende — APPROVED** [review]

**Pull request merged**